### ISSUE TITLE
Issue with generic function type passed into object. 

### ISSUE THREAD

**Issue by rostislav-simonik** [microsoft/TypeScript#40403]
**TypeScript Version:**  4.0.2

**Search Terms:**

**Code**

```ts
type Identity = <TYPE>(attributes: TYPE) => TYPE

export const map: { [key: string]: Identity } = {
  add: (attributes: string) => attributes,
}
```

**Expected behavior:**
Should not throw an error

**Actual behavior:**
throwing 
```
Type '(attributes: string) => string' is not assignable to type 'Identity'. Types of parameters 'attributes' and 'attributes' are incompatible. Type 'TYPE' is not assignable to type 'string'.
```

**Playground Link:** 
https://www.typescriptlang.org/play?#code/C4TwDgpgBAkgJhAdsAlqKBeKAeAKgTQAUBRAPgAoBDYYAJxQCMBXYCAZwC4oCSBKTUtyLEAUCIgAPMAHtawKAGNpiNvIC2lMFwDeUANoBrCCC6r6iAOYBdLvCSp0AX0xRtIqFEpw4XKjXrMrJxQZiiW-BiC1HSMLOwANCKOQA

**Comment by MartinJohns** [comment]
You're not passing a generic function, and that is the actual problem. Your type `{[key:string]:Identity}` says that every property of your object stores a value matching the `Identity` type, which in turn is a generic function. However, the value you pass in is **not** a generic function, but instead a function accepting and returning `string`. This can't work.

**Comment by MartinJohns** [comment]
Here is an example why your code is fundamentally flawed:

    type Identity = <TYPE>(attributes: TYPE) => TYPE

    export const map: { [key: string]: Identity } = {
      // @ts-ignore
      // Returning a string here, not the input argument.
      add: (attributes: string) => "hello example",
    }

    // The type is "Identity", so a generic function.
    const addIdentity: Identity = map["add"];

    // Calling the generic function with number as the type argument.
    const result = addIdentity<number>(5);

    // According to the types it should be a number, but it's actually a string.
    console.log(typeof result);

**Comment by rostislav-simonik** [comment]
@MartinJohns thx, for response.

yes your example is right, i wanna use it similar way and retrieve function later. 
Important to me is validation when function is added to map, so attributes type is the same as return type. 

is there alternative way how to achieve that, not using generic function type?

**Comment by RyanCavanaugh** [comment]
More examples of how you wanted to use this would help, but this isn't a bug

**Comment by rostislav-simonik** [comment]
@RyanCavanaugh thx, for answer. I met problem again, sorry i somehow missed your last message. 

```

// NOTE: i want to enforce that record will contain functions that will have same type on return type as argument type 
// but for each handler given type can be different 
type B = <A>(a: A) => A

const c: Record<string, B> = {
    num: (a: number): number => a, // DESIRED
    str: (a: string): string => a, // DESIRED
    mix: (a: number): string => a.toString(), // NOT DESIRED
}
```
https://www.typescriptlang.org/play?ts=4.2.3#code/FAehAIDkHkBUFEBc4CW4DuBDAdgF3LgPbgCm2AZoQE4DGJBAFpvlSTdQCYYoA2P47PJhTZw5AK7YauFIWwBnRs259wTAG715mALb1cATwAO9OeFa5xVUYZPhMizFQDm4vXgLH6oCACNx+JRUpJg0DGo4HDwkwc4omjZeAjjgvvQcKOTkMWT4wLb0AELgALzgADwAggB8ABSYyJUAlKXV4JXAwILy+DTIAEpsnOU9VCLOADTghW1lAN7A4Evg2G7I9cirOmlUTZtuO632U2DgACLwAMoAkv3wZ4vLo+sN4KPje2+4Y9jOR5gnCAXG53B7LcA6FAADxe+22MU+71+-wAdERLt9xrUmoCoHBzldbvdgABfIA

**Comment by RyanCavanaugh** [comment]
@rostislav-simonik there's not really a way to do that. Inference of type parameters does not occur during simple assignment, and even if it did, there's not anything like per-property type parameter inference.

**Comment by rostislav-simonik-plc** [comment]
Ok thx for confirmation, i'll just make them `any`, 

just wondering if we could be achive in future with the nominal proposal and then to have something like this 

https://github.com/Microsoft/TypeScript/issues/202

```

nominal A = any;  

type B = (a: A) => A

const c: Record<string, B> = {
    num: (a: number): number => a, //ok
    str: (a: string): string => a, // ok
    mix: (a: number): string => a.toString(), // throws error
}
```
